### PR TITLE
Relies on the Storage trait instead of tikv::storage

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

Extracted from https://github.com/tikv/tikv/pull/5119

This PR makes Coprocessor DAG completely does not rely on tikv::storage (including MVCC scanners and Statistics).

The next step will be make Coprocessor DAG does not rely on tikv::coprocessor (especially coprocessor::Error), so that Coprocessor DAG can be a standalone workspace and anything connects to TiKV is tikv::coprocessor.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

No need.

## Refer to a related PR or issue link (optional)

See https://github.com/tikv/tikv/issues/4180 Extract Coprocessor DAG

### CODE REVIEW

File: benches/coprocessor_executors/hash_aggr/util.rs [hunk 1]
```diff
@@ -63,7 +64,7 @@ impl HashAggrBencher for NormalBencher {
                 black_box(Box::new(src)),
             )
             .unwrap();
-            Box::new(ex) as Box<dyn Executor>
+            Box::new(ex) as Box<dyn Executor<StorageStats = Statistics>>
```
Comment (sticnarf) [inline benches/coprocessor_executors/hash_aggr/util.rs]: In the benches, can we use `()` as the stat type? It is real pain to compile benches. We don't want to depend on the tikv trait to run benches.

Comment (breeswish) [reply benches/coprocessor_executors/hash_aggr/util.rs]: Good suggestion. I will refine it.

Comment (breeswish) [reply benches/coprocessor_executors/hash_aggr/util.rs]: Considering that this benchmark itself also benches *real RocksDB scenario*, would you accept a feature gate solution?

Comment (sticnarf) [reply benches/coprocessor_executors/hash_aggr/util.rs]: I find it easy to extract the RocksDB engine out.
(Actually, I find `storage::Statistics` seems easy to be extracted out too...If we extract storage statistics, it seems not so useful that we split coprocessor execution and storage statistics...)

Comment (breeswish) [reply benches/coprocessor_executors/hash_aggr/util.rs]: @sticnarf As a standalone query executor, it should not depend on specific storage statistics.

Comment (sticnarf) [reply benches/coprocessor_executors/hash_aggr/util.rs]: Makes sense.

File: src/coprocessor/dag/builder.rs [hunk 2]
```diff
@@ -150,61 +152,78 @@ impl DAGBuilder {
...
-                        .with_summary_collector(C::new(summary_slot_index)),
-                    )
-                }
-                ExecType::TypeAggregation | ExecType::TypeStreamAgg
-                    if ed.get_aggregation().get_group_by().is_empty() =>
-                {
-                    COPR_EXECUTOR_COUNT
-                        .with_label_values(&["batch_simple_aggr"])
-                        .inc();
+            let new_executor: Box<dyn BatchExecutor<StorageStats = S::Statistics>> =
```
Comment (sticnarf) [inline src/coprocessor/dag/builder.rs]: Quite annoying diff caused by reformatting :(

File: src/coprocessor/dag/executor/mod.rs [hunk 3]
```diff
@@ -387,110 +385,22 @@ pub mod tests {
...
-        table_scan.set_columns(cis.clone().into());
-
-        let key_ranges = key_ranges.unwrap_or_else(|| vec![get_range(tid, 0, i64::max_value())]);
-
-        let (snapshot, start_ts) = test_store.get_snapshot();
-        let store = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, true);
-        Box::new(TableScanExecutor::table_scan(table_scan, key_ranges, store, false).unwrap())
+    #[inline]
+    pub fn get_range(table_id: i64, start: i64, end: i64) -> KeyRange {
+        let mut key_range = KeyRange::default();
```
Comment (lonng) [inline src/coprocessor/dag/executor/mod.rs]: Changing `KeyRange::default` to `KeyRange::new` in L392 (https://github.com/tikv/tikv/pull/5143/files#diff-afaa9539aadbecbcf6c9af032e38da84R392), but use `KeyRange::default` here, is there special usage?

File: src/coprocessor/dag/builder.rs (hunk 2 above)
Comment (lonng) [reply src/coprocessor/dag/builder.rs]: Ues `Split` mode to review this file will make your life better.

File: src/coprocessor/dag/executor/mod.rs (hunk 3 above)
Comment (breeswish) [reply src/coprocessor/dag/executor/mod.rs]: Sorry this should be a merge mistake. It should always be `::default()`. See https://github.com/tikv/tikv/commit/f5215d631d3145d824c0d0940cd490ff3f8bbf80

**Pull request merged**